lib/fluent/plugin/filter_add_insert_ids.rb

```diff
@@ -0,0 +1,88 @@
+# Copyright 2018 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'fluent/plugin/filter'
+require 'thread'
+
+module Fluent
+  module Plugin
+    # Fluentd filter plugin for adding insertIds to guarantee log entry order
+    # and uniqueness.
+    class AddInsertIdsFilter < Filter
+      Fluent::Plugin.register_filter('add_insert_ids', self)
+
+      # Constants for configuration.
+      module ConfigConstants
+        # The default field name of insertIds in the log entry.
+        DEFAULT_INSERT_ID_KEY = 'logging.googleapis.com/insertId'.freeze
+        # The character size of the insertIds. This matches the setup in the
+        # Stackdriver Logging backend.
+        INSERT_ID_SIZE = 17
+        # The characters that are allowed in the insertIds. This matches the
+        # allowed collection by the Stackdriver Logging Backend.
+        ALLOWED_CHARS = (Array(0..9) + Array('a'..'z')).freeze
+      end
+
+      include self::ConfigConstants
+
+      desc 'The field name of insertIds in the log entry.'
+      config_param :insert_id_key, :string, default: DEFAULT_INSERT_ID_KEY
+
+      # Expose attr_readers for testing.
+      attr_reader :insert_id_key
+
+      def start
+        super
+        @log = $log # rubocop:disable Style/GlobalVars
+
+        # Initiate the insertID.
+        @log.info "Started the add_insert_ids plugin with #{@insert_id_key}" \
+                  ' as the insert ID key.'
+        @insert_id = generate_insert_id
+        @log.info "Initiated the insert ID key at #{@insert_id}."
+      end
+
+      def configure(conf)
+        super
+      end
+
+      def shutdown
+        super
+      end
+
+      # rubocop:disable Style/UnusedMethodArgument
+      def filter(tag, time, record)
+        # Only generate and add an insertId field If the record is a hash and
+        # the insert ID field is not already set.
+        if record.is_a?(Hash) && !record.key?(@insert_id_key)
```

Comment:
Should we check for empty values as well?

Reply:
`is_a?` checks that already.

```
2.4.3 :004 > ''.is_a?(Hash)
 => false
2.4.3 :005 > nil.is_a?(Hash)
 => false
```

Reply:
Synced offline, checking the actual value within the record to make sure it's not blank. It might be over-optimizing, if the user really wants blank, that would be strange, but maybe it's intentional? I'll let you make the judgement call here as to whether or not blank values are an important scenario to auto-generate ids.

Reply:
Added a check against an empty string.